storage/innobase/trx/trx0rec.cc

```diff
@@ -2038,170 +2038,117 @@ trx_undo_report_row_operation(
+@param v_heap    memory heap used to create vrow dtuple if it is not yet
+                 created. This heap diffs from "heap" above in that it could be
+                 prebuilt->old_vers_heap for selection
+@param vrow      virtual column info, if any
 @return error code
 @retval DB_SUCCESS if previous version was successfully built,
 or if it was an insert or the undo record refers to the table before rebuild
 @retval DB_MISSING_HISTORY if the history is missing */
 TRANSACTIONAL_TARGET
-dberr_t
-trx_undo_prev_version_build(
-	const rec_t 	*rec,
-	dict_index_t	*index,
-	rec_offs	*offsets,
-	mem_heap_t	*heap,
-	rec_t		**old_vers,
-	mem_heap_t	*v_heap,
-	dtuple_t	**vrow,
-	ulint		v_status)
+dberr_t trx_undo_prev_version_build(const rec_t *rec, dict_index_t *index,
+                                    rec_offs *offsets, mem_heap_t *heap,
+                                    rec_t **old_vers, mtr_t *mtr,
+                                    ulint v_status,
+                                    mem_heap_t *v_heap, dtuple_t **vrow)
 {
-	dtuple_t*	entry;
-	trx_id_t	rec_trx_id;
-	undo_no_t	undo_no;
-	table_id_t	table_id;
-	trx_id_t	trx_id;
-	roll_ptr_t	roll_ptr;
-	upd_t*		update;
-	byte		type;
-	byte		info_bits;
-	byte		cmpl_info;
-	bool		dummy_extern;
-	byte*		buf;
+  ut_ad(!index->table->is_temporary());
+  ut_ad(rec_offs_validate(rec, index, offsets));
 
-	ut_ad(!index->table->is_temporary());
-	ut_ad(rec_offs_validate(rec, index, offsets));
+  const roll_ptr_t roll_ptr= row_get_rec_roll_ptr(rec, index, offsets);
+  *old_vers= nullptr;
 
-	roll_ptr = row_get_rec_roll_ptr(rec, index, offsets);
+  if (trx_undo_roll_ptr_is_insert(roll_ptr))
+    /* The record rec is the first inserted version */
+    return DB_SUCCESS;
 
-	*old_vers = NULL;
+  ut_ad(roll_ptr < 1ULL << 55);
+  ut_ad(uint16_t(roll_ptr) >= TRX_UNDO_PAGE_HDR + TRX_UNDO_PAGE_HDR_SIZE);
+  ut_ad(uint32_t(roll_ptr >> 16) >= FSP_FIRST_INODE_PAGE_NO);
 
-	if (trx_undo_roll_ptr_is_insert(roll_ptr)) {
-		/* The record rec is the first inserted version */
-		return DB_SUCCESS;
-	}
+  const trx_id_t rec_trx_id= row_get_rec_trx_id(rec, index, offsets);
 
-	mariadb_increment_undo_records_read();
-	rec_trx_id = row_get_rec_trx_id(rec, index, offsets);
+  ut_ad(!index->table->skip_alter_undo);
 
-	ut_ad(!index->table->skip_alter_undo);
+  mariadb_increment_undo_records_read();
+  const auto savepoint= mtr->get_savepoint();
+  dberr_t err= DB_MISSING_HISTORY;
+  purge_sys_t::view_guard check{v_status == TRX_UNDO_CHECK_PURGE_PAGES
+                                ? 0
+                                : v_status == TRX_UNDO_CHECK_PURGEABILITY
+                                ? 1 : -1};
+  if (!check.view().changes_visible(rec_trx_id))
+  {
+    trx_undo_rec_t *undo_rec= nullptr;
+    if (const buf_block_t *undo_page=
+        check.get(page_id_t{trx_sys.rseg_array[(roll_ptr >> 48) & 0x7f].
+                            space->id,
+                            uint32_t(roll_ptr >> 16)}, mtr))
+    {
```

Comment:
The code from `trx_undo_decode_roll_ptr`() abstraction has come out. In general it is best to keep such decode rules in one place especially with so many non-intuitive constants like 48, 16, 0x7f etc.

Reply:
yes, I know. A wrapper would be good.